### PR TITLE
Add location to shops.

### PR DESCRIPTION
**Pull request by starkshaw**

### CODE REVIEW

File: app/views/shops/_form.html.haml [hunk 1]
```diff
@@ -25,4 +25,23 @@
...
 			= f.email_field :email_id
 			%br/
-			= f.submit(class: "btn btn-primary")
\ No newline at end of file
+			Location:
+			%input#address{:style => "width: 200px", :type => "text"}/
+			#map{:style => "width: 500px; height: 400px;"}
+			= f.hidden_field :latitude
+			= f.hidden_field :longitude
+			:javascript
```
Comment (danielpaul) [inline app/views/shops/_form.html.haml]: Could we move this to an external JS file that would be included in the application.js? Need to have jQuery ready function in that and an if condition to check if the map ID exists, then add the javascript here inside that. Having inline javascript is not best practices. In the application.js they will all be compressed and minified, etc.

Comment (starkshaw) [inline app/views/shops/_form.html.haml]: I don't know much about it, need help.

File: db/migrate/20160226103946_add_location_to_shop.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class AddLocationToShop < ActiveRecord::Migration
+  def change
+    add_column :shops, :longitude, :numeric
```
Comment (danielpaul) [inline db/migrate/20160226103946_add_location_to_shop.rb:3]: Does numeric accept decimal points and negative numbers? Don't we need decimal to save decimal points in the coordinates? Also index the two columns for faster queries. 

```
    add_column :shops, : longitude, :decimal
    add_column :shops, : latitude, :decimal

    add_index  :shops, [:lat, :lng]
```

File: app/views/shops/_form.html.haml (hunk 1 above)
Comment (danielpaul) [inline app/views/shops/_form.html.haml]: `map.js` file something like this:

```

$(function() {

  var load_map = function() {

    $('#map').locationpicker({
        // location: {latitude: 53.3482993, longitude: -6.257899},
        location: {latitude: $("#address_lat").val() || 53.3482993, longitude: $('#address_lng').val() || -6.257899},
        inputBinding: {
          latitudeInput: $('#address_lat'),
          longitudeInput: $('#address_lng'),
          locationNameInput: $('#address_lookup')
        },
        enableAutocomplete: true,
        radius: 0,
        zoom: 16,
        scrollwheel: true
    });
  }

  if($('#map').length) load_map();

});

```

You need to update all the variable names and IDs in the code above if you are using it exactly. But just use the above as a template for the code you already have.

File: db/migrate/20160226103946_add_location_to_shop.rb (hunk 2 above)
Comment (starkshaw) [inline db/migrate/20160226103946_add_location_to_shop.rb:3]: Numeric works fine in database too **AS I STATED BEFORE**.

Index added.

File: db/migrate/20160226103946_add_location_to_shop.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+    add_column :shops, :latitude, :numeric
+    add_index  :shops, [:lat, :lng]
```
Comment (danielpaul) [inline db/migrate/20160226103946_add_location_to_shop.rb]: Sorry, I forgot to change the names of the columns in my copy-paste code above, but you need to change `lat` → `latitude` & `lng` → `longitude`.

Comment (danielpaul) [inline db/migrate/20160226103946_add_location_to_shop.rb]: And also need to run `rake db:migrate` to add the index to `schema.rb`.

File: app/views/shops/_form.html.haml (hunk 1 above)
Comment (starkshaw) [inline app/views/shops/_form.html.haml]:
```
Location:
%input#address{:style => "width: 200px", :type => "text"}/
#map{:style => "width: 500px; height: 400px;"}
= f.hidden_field :latitude
= f.hidden_field :longitude
:javascript
    var lat = parseFloat($('#shop_latitude').val())
    var lon = parseFloat($('#shop_longitude').val())
    $('#map').locationpicker({
        location: {latitude: lat || 53.348864, longitude: lon || -6.2502467},
        radius: 0,
        enableAutocomplete: true,
        inputBinding: {
            latitudeInput: $('#shop_latitude'),
            longitudeInput: $('#shop_longitude'),
            locationNameInput: $('#address')
        }
    });
%br/
= f.submit(class: "btn btn-primary")
```

Check the code above, the reason I didn't save it as an external file is the rails `id`, such as `:latitude`, `:latitude` is differed based on the view. I need to hard code them in the JavaScript, that's why in JS part the `id` are `#shop_latitude` and `#shop_longitude`.

Comment (danielpaul) [inline app/views/shops/_form.html.haml]: You can overwrite the field's ID. 

`= f.hidden_field :latitude, id: 'latitude'`

Or if you want to make it more dynamic, have a `#map_wrapper` and in your JS code you find the lat and lng hidden fields within that. But do the first option is much easier and clearer.

**Comment by starkshaw** [comment]
Finished.

**Pull request merged**